Svc/FpySequencer/FpySequencer.cpp

```diff
@@ -0,0 +1,268 @@
+    if (sequencer_getState() != State::AWAITING_CMD_RUN_VALIDATED) {
+        this->log_WARNING_HI_InvalidCommand(static_cast<I32>(sequencer_getState()));
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+        return;
+    }
+
+    if (block == FpySequencer_BlockState::BLOCK) {
+        // save the opCode and cmdSeq so we can respond later
+        this->m_savedOpCode = opCode;
+        this->m_savedCmdSeq = cmdSeq;
+    }
+
+    this->sequencer_sendSignal_cmd_RUN_VALIDATED(FpySequencer_SequenceExecutionArgs(this->m_sequenceFilePath, block));
+
+    // only respond if the user doesn't want us to block further execution
+    if (block == FpySequencer_BlockState::NO_BLOCK) {
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+    }
+}
+
+//! Handler for command CANCEL
+//!
+//! Cancels a running or validated sequence
+void FpySequencer::CANCEL_cmdHandler(FwOpcodeType opCode,  //!< The opcode
+                                     U32 cmdSeq            //!< The command sequence number
+) {
+    // only state you can't cancel in is IDLE
+    if (sequencer_getState() == State::IDLE) {
+        this->log_WARNING_HI_InvalidCommand(static_cast<I32>(sequencer_getState()));
+        this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::EXECUTION_ERROR);
+        return;
+    }
+
+    this->sequencer_sendSignal_cmd_CANCEL();
+
+    // cancel returns immediately and always succeeds
+    this->cmdResponse_out(opCode, cmdSeq, Fw::CmdResponse::OK);
+}
+
+//! Handler for input port checkTimers
+void FpySequencer::checkTimers_handler(FwIndexType portNum,  //!< The port number
+                                       U32 context           //!< The call order
+) {
+    this->sequencer_sendSignal_checkTimersIn();
+}
+
+void FpySequencer::pingIn_handler(FwIndexType portNum, /*!< The port number*/
+                                  U32 key              /*!< Value to return to pinger*/
+) {
+    // send ping response
+    this->pingOut_out(0, key);
+}
+
+//! Handler for input port cmdResponseIn
+void FpySequencer::cmdResponseIn_handler(FwIndexType portNum,             //!< The port number
+                                         FwOpcodeType opCode,             //!< Command Op Code
+                                         U32 cmdSeq,                      //!< Command Sequence
+                                         const Fw::CmdResponse& response  //!< The command response argument
+) {
+    // TODO ask Rob if there's a better way to check if we're in a superstate. I don't want to have
+    // to update this every time I add a new substate to the RUNNING state.
+
+    // if we aren't in the RUNNING state:
+    if (this->sequencer_getState() != State::RUNNING_AWAITING_STATEMENT_RESPONSE
+        && this->sequencer_getState() != State::RUNNING_DISPATCH_STATEMENT
+        && this->sequencer_getState() != State::RUNNING_SLEEPING) {
+        // must be a coding error.
```

Comment:
Must be a coding error or a canceled sequence response.

Reply:
Perhaps this should also be "CmdFromOldSequence"

Reply:
I am going to resolve this.  It is a superset of the event type lacking specific parameters.